rialto_airflow/harvest/sul_pub.py

```diff
@@ -1,53 +1,91 @@
 
-SUL_PUB_FIELDS = [
-    "authorship",
-    "title",
-    "author",
-    "year",
-    "type",
-    "journal",
-    "provenance",
-    "doi",
-    "sulpubid",
-    "date",
-    "country",
-    "booktitle",
-]
-
-
-def sul_pub_csv(csv_file, host, key, since=None, limit=None):
-    with open(csv_file, "w") as csvfile:
-        writer = csv.DictWriter(csvfile, fieldnames=SUL_PUB_FIELDS)
-        writer.writeheader()
-        for row in harvest(host, key, since, limit):
-            writer.writerow(row)
-
-
-def harvest(host, key, since, limit):
+
+def harvest(snapshot, host, key, per_page=1000, limit=None):
+    # create a jsonl file to write to
+    jsonl_file = snapshot.path / "sulpub.jsonl"
+
+    with jsonl_file.open("w") as jsonl_output:
+        for sulpub_pub in publications(host, key, per_page, limit):
+            with get_session(snapshot.database_name).begin() as session:
+                doi = extract_doi(sulpub_pub)
+
+                # if when inserting the row we get a constraint violation
+                # on the DOI then we have to do an update
+                pub_id = session.execute(
+                    insert(Publication)
+                    .values(doi=doi, sulpub_json=sulpub_pub)
+                    .on_conflict_do_update(
+                        constraint="publication_doi_key",
+                        set_=dict(sulpub_json=sulpub_pub),
+                    )
+                    .returning(Publication.id)
+                ).scalar_one()
+
+                # get a list of approved cap_ids for the publication
+                # the cap_profile_id needs to be a string for the database
+                cap_ids = [
+                    str(a["cap_profile_id"])
+                    for a in sulpub_pub["authorship"]
+                    if a["status"] == "approved"
+                ]
+
+                # if the row is already there we could get a constraint
+                # violation, but we can safely ignore that
+                session.execute(
+                    insert(pub_author_association)
+                    .from_select(
+                        ["publication_id", "author_id"],
+                        select(pub_id, Author.id).where(
+                            Author.cap_profile_id.in_(cap_ids)
+                        ),
+                    )
+                    .on_conflict_do_nothing()
+                )
+
+                jsonl_output.write(json.dumps(sulpub_pub) + "\n")
+
+    return jsonl_file
+
+
+def publications(host, key, per_page=1000, limit=None):
     url = f"https://{host}/publications.json"
 
     http_headers = {"CAPKEY": key}
-
-    params = {"per": 1000}
-    if since:
-        params["changedSince"] = since.strftime("%Y-%m-%d")
+    params = {"per": per_page}
 
     page = 0
     record_count = 0
     more = True
 
+    # catch and back off from temporary errors from sulpub
+    # https://docs.python-requests.org/en/latest/user/advanced/#example-automatic-retries
+
```

Comment:
Try to catch and retry connection errors from sulpub. I saw one when running Airflow at home on my laptop.